### PR TITLE
Fix issues related to layouts

### LINKED ISSUE #2357: Inconsistencies while changing between panel layouts
Steps to reproduce the issue:
1. Run ZAP with add-ons "Quick Start" and "Script Console" installed (unpin and close its tabs, if not already);
2. While in "Maximise left (Sites) tab" or "Maximise bottom (History etc) tabs" layout switch to "Full Layout" (Options > Display > "Display:" or using the main tool bar options), note the tabs currently visible;
3. Change back to "Maximise left (Sites) tab" or "Maximise bottom (History etc) tabs", note that tabs that were hidden are now shown (for example, "Quick Start" tab), moreover the plus sign is shown with tabs that are visible and that some of the tabs belong to other panels (for example, "Quick Start", "Break", "Scripts" tabs are available in the footer panel).

ZAP Versions:
Versions >= 2.4.0

### LINKED ISSUE #2358: ZAP fails to start in "Full Layout"
Issue reported in OWASP ZAP Developer Group: https://groups.google.com/d/topic/zaproxy-develop/i5nqsL_RYs4/discussion

Steps to reproduce the issue:
1. Run ZAP;
2. Change layout to "Full Layout" (Options > Display > "Display:" or using the main toolbar option "Expand Full")
3. Restart ZAP and note that it does not start;
4. Check the log file, there should be an exception:
```
[AWT-EventQueue-0] INFO org.parosproxy.paros.view.View  - Initialising View
[AWT-EventQueue-0] ERROR org.zaproxy.zap.ZAP$UncaughtExceptionLogger  - Exception in thread "AWT-EventQueue-0"
java.lang.NullPointerException
	at org.parosproxy.paros.view.View.getWorkbench(View.java:884)
	at org.parosproxy.paros.view.View.getMessagePanelsPositionController(View.java:327)
	at org.parosproxy.paros.view.WorkbenchPanel.initialize(WorkbenchPanel.java:188)
	at org.parosproxy.paros.view.WorkbenchPanel.changeDisplayOption(WorkbenchPanel.java:249)
	at org.parosproxy.paros.view.MainFrame.changeDisplayOption(MainFrame.java:178)
	at org.parosproxy.paros.view.MainFrame.<init>(MainFrame.java:69)
	at org.parosproxy.paros.view.View.init(View.java:227)
	at org.parosproxy.paros.view.View.getSingleton(View.java:581)
	at org.zaproxy.zap.GuiBootstrap.init(GuiBootstrap.java:135)
	at org.zaproxy.zap.GuiBootstrap.startImpl(GuiBootstrap.java:103)
	at org.zaproxy.zap.GuiBootstrap.access$0(GuiBootstrap.java:87)
	at org.zaproxy.zap.GuiBootstrap$1.run(GuiBootstrap.java:81)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:312)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:745)
	at java.awt.EventQueue.access$300(EventQueue.java:103)
	at java.awt.EventQueue$3.run(EventQueue.java:706)
	at java.awt.EventQueue$3.run(EventQueue.java:704)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:715)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:242)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:161)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:150)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:146)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:138)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:91)
```

Remarks:
Issue introduced in #2121 (61dab8de452dbe445e956484c085d4ec5384cdd0).

ZAP Versions:
Weekly releases >= 2015-12-14

### LINKED ISSUE #2359: Show Tab Names option not applied to Response tab
Steps to reproduce the issue:
1. Run ZAP;
2. Change position of "Response" tab to "Request shown above Response" or "Request and Response side by side" (using tool bar buttons);
3. Set/Unset the option "Show Tab Names" several times and note that the changes are not immediately applied to "Response" tab (only after changing "Response" tab position again).

ZAP Versions:
Versions >= 2.4.0

### LINKED ISSUE #2373: Show Tab not working correctly in Full Layout for non-information tabs
Steps to reproduce the issue:
1. Run ZAP with add-on "Quick Start" installed (unpin and close its tab, if not already);
2. Switch to Full Layout (Options > Display > "Display:" or using the main tool bar option);
3. Attempt to show the "Quick Start" tab using the menu item "View" > "Show Tab" > "Quick Start Tab" (or keyboard shortcut Ctrl + Shift + Q) and note that the tab is not shown.

Remarks:
In Full Layout only the information/footer tabs are shown when using the menus/keyboard shortcuts, unless previously shown using the green plus menu.

ZAP Versions:
Versions >= 2.4.0

### LINKED ISSUE #2374: Unable to change response tab position without main tool bar
Steps to reproduce the issue:
1. Start ZAP;
2. Deselect the option "Options" > "Display" > "Show Main Toolbar" and restart ZAP;
3. Note that it's no longer possible to change the position of the response tab, the option is not available in the "Display" panel of the Options dialogue (nor other place).

ZAP Versions:
Versions >= 2.1.0

### PR DESCRIPTION
**Pull request by thc202**
Move the layout management to class WorkbenchPanel which has all the
information necessary to keep its layouts/options in the expected state.

Following more detailed explanation of the changes done to each class:
 - MenuToolsControl, notify the MainFrame class that the options were
 changed, to apply the view options;
 - ExtensionLoader, remove direct manipulation of the tabbed panels of
 WorkbenchPanel class and remove code related to layout logic, the
 addition and removal of panels are done through methods of
 WorkbenchPanel class;
 - OptionsParamView:
   - Add method to read and to persist the position of the response
   panel (logic/code moved from MessagePanelsPositionController);
   - Remove the access/change of the WorkbenchPanel when Show Tab Names
   option is changed, that's now done by MainFrame when applying the
   view options;
 - OptionsViewPanel, add a new option to allow to change the position of
 the response panel, so that's also available when the main tool bar is
 not visible;
 - MainFrame:
   - Move the layout buttons from MainToolbarPanel to this class, to
   avoid the circular dependency between MainToolbarPanel and MainFrame,
   with this change MainToolbarPanel no longer needs to call the
   MainFrame when the options are changed. Also, moved the response
   panel position buttons from class MessagePanelsPositionController.
   MainFrame adds the buttons to the MainToolbarPanel when created and
   takes care to set the buttons enabled when the options are changed;
   - Add a method that applies the view options, visibility of the main
   tool bar, the layout and response panel position of the workbench
   panel and if the tabs should display the panels' names. It takes care
   to set the state of the options to WorkbenchPanel and the state of
   the buttons added to the main tool bar;
   - Add other methods that control the options previously mentioned, to
   keep the state of the main tool bar buttons and workbench panel in
   sync.
 - TabbedPanel, change to use ComponentMaximiser and
 ComponentMaximiserMouseListener;
 - View:
   - Change to use the new methods of MainFrame and WorkbenchPanel,
   instead of accessing/manipulating  directly the components of the
   WorkbenchPanel class;
   - Deprecated methods and constants used to change the layout of the
   WorkbenchPanel, that's done through the MainFrame class;
 - WorkbenchPanel:
   - Move all layout logic done in MessagePanelsPositionController to
   the class, to have, in all cases, the correct layout/options applied;
   - Add a JavaDoc comment to discourage direct manipulation of the
   tabbed panels, the changes might be lost when changing layouts;
   - Expose helper methods to change the state of the (tabbed) panels,
   called by MainFrame and View, instead of changing them directly;
   - Add enum PanelType to define the type of the panels added,
   replaces, along with add/remove methods, the direct manipulation of
   the tabbed panels;
   - Add enums Layout and ResponsePanelPosition, to define the layouts
   and response panel positions supported;
 - GuiBootstrap, remove the statement that was setting the layout to the
 View, no longer needed, MainFrame already takes care to read and apply
 its (and child components) view options;
 - ComponentMaximiser, extracted from TabbedPanel, helper class to
 maximise a component;
 - ComponentMaximiserMouseListener, extracted from TabbedPanel, helper
 class to maximise a component when the component is clicked twice;
 - MainToolbarPanel:
   - As mentioned previously, some of the buttons were moved to
   MainFrame class;
   - Deprecate method that set the layout, no longer used;
 - MessagePanelsPositionController, remove all code and deprecate the
 class, the position of the response panel is done by WorkbenchPanel
 class.
 - TabbedPanel2:
   - Deprecate the method clone(TabbedPanel2), no longer used and the
   implementation is not correct;
   - Change method setVisible(Component, boolean) to keep the hideable
   state of the panel when adding it, otherwise it would be possible to
   close it even if it shouldn't;
   - Change method addTab(String, Icon, Component, boolean, boolean,
   int):
     - To add the panel to the list of removed panels if it's visible,
     otherwise it would be shown in the green plus menu even if visible;
     - Ensure that a tab is always selected, client code no longer needs
     to do that;
   - Override removeAll() to update the internal list that tracks which
   panels are visible and to update the green plus menu;
   - Add methods to obtain and set the visible panels and obtain all
   panels, so that the visible state of the panels can be applied to
   other layouts.

Fix #2357 - Inconsistencies while changing between panel layouts
Fix #2358 - ZAP fails to start in "Full Layout"
Fix #2359 - Show Tab Names option not applied to Response tab
Fix #2373 - Show Tab not working correctly in Full Layout for
non-information tabs
Fix #2374 - Unable to change response tab position without main tool bar

### CODE REVIEW

File: src/org/parosproxy/paros/view/MainFrame.java [hunk 1]
```diff
@@ -26,48 +26,201 @@
...
+	private final WorkbenchPanel paneStandard;
 	private org.parosproxy.paros.view.MainMenuBar mainMenuBar = null;
 	private JPanel paneDisplay = null;
 
 	private MainToolbarPanel mainToolbarPanel = null;
 	private MainFooterPanel mainFooterPanel = null;
 
-	private int displayOption;
+	/**
+	 * The {@code ZapToggleButton} that sets whether or not the tabs should shown the panels' names.
```
Comment (kingthorin) [inline src/org/parosproxy/paros/view/MainFrame.java]: Should be "show" not "shown"....

File: src/org/parosproxy/paros/view/MainFrame.java [hunk 2]
```diff
@@ -110,16 +263,12 @@ private JPanel getPaneContent() {
 	}
 
 	/**
-	 * This method initializes paneStandard
+	 * Gets the {@code WorkbenchPanel}.
 	 *
-	 * @return org.parosproxy.paros.view.StandardPanel
+	 * @return the workbench panel
+	 * @since 2.2.0
```
Comment (kingthorin) [inline src/org/parosproxy/paros/view/MainFrame.java]: Huh?

File: src/org/zaproxy/zap/view/ComponentMaximiserMouseListener.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+     * @return the {@code ComponentMaximiser} that's used to maximise the components, might be {@code null}
+     * @see #setComponentMaximiser(ComponentMaximiser)
+     */
+    public ComponentMaximiser getComponentMaximiser() {
+        return componentMaximiser;
+    }
+
+    /**
+     *
+     */
```
Comment (kingthorin) [inline src/org/zaproxy/zap/view/ComponentMaximiserMouseListener.java]: Drop the empty javadoc?

File: src/org/parosproxy/paros/view/MainFrame.java (hunk 1 above)
Comment (thc202) [inline src/org/parosproxy/paros/view/MainFrame.java]: Done.

File: src/org/zaproxy/zap/view/ComponentMaximiserMouseListener.java (hunk 3 above)
Comment (thc202) [inline src/org/zaproxy/zap/view/ComponentMaximiserMouseListener.java]: Right, that was the idea. Done.

**Comment by kingthorin** [comment]
Looks good

**Pull request merged**